Allow QueueManager to operate in parallel

milljm opened the pull request:
Minimum required to get QueueManager to work with threads.

Allow the QueueManager to utilize the threading model provided in
the base class.

Added methods to the job class to; 1. return a unique identifier,
and *2. obtain/set the number of processors needed for that job.
Modifications to the Scheduler base class were made to use this
new method of resource allocation.

* This is necessary when using the QueueManager, because each job
launched requires only 1 processor in order to do so (invoking
qsub, qstat, etc), and _not_ how many processors the _test_ needs.

Modified contrib/dag to allow the retrieval of a cached dag, and
added methods to reverse the edges in the dag so the QueueManager
can reverse the dependency direction when performing
processResults.

Fixed a bug where I was clobbering the tester object while in a for
loop during the race condition checks.

Fix a bug when supplying absolute path with --pbs options, the path
was stipped.

Add unit tests for RunPBS

Closes #9911

Commit:
Allow QueueManager to operate in parallel

Allow the QueueManager to utilize the threading model provided in
the base class.

Added methods to the job class to; 1. return a unique identifier,
and *2. obtain/set the number of processors needed for that job.
Modifications to the Scheduler base class were made to use this
new method of resource allocation.

* This is necessary when using the QueueManager, because each job
launched requires only 1 processor in order to do so (invoking
qsub, qstat, etc), and _not_ how many processors the _test_ needs.

Modified contrib/dag to allow the retrieval of a cached dag, and
added methods to reverse the edges in the dag so the QueueManager
can reverse the dependency direction when performing
processResults.

Fixed a bug where I was clobbering the tester object while in a for
loop during the race condition checks.

Fix a bug when supplying absolute path with --pbs options, the path
was stipped.

Add unit tests for RunPBS

Closes #9931

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/123703/) on afd4b94 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10046/site)

This comment will be updated on new commits.

milljm commented on python/TestHarness/tests/test_QueuePBS.py:
```diff
@@ -0,0 +1,418 @@
+
+                # but our second job should have copied some extra stuff over
+                elif job_id == 1 and not os.path.exists(os.path.join(self.session_dir, 'queue_job_copy_file_test')):
+                    raise Exception('Failed: A file was not copied that should have been')
```
Should probably use what ever file is listed in the test file instead of arbitrarily entering it here. And perhaps test multiple files.

```python
    for file in specs['copy_files']:
```

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -533,9 +555,12 @@ def runWorker(self, job_container):
                 next_job_group = self.getNextJobGroup(job_dag)
 
+                # Allow derived schedulers to perform post run operations
+                self.postRun(job_container)
```
This should run _after_ getNextJobGroup, because getNextJobGroup deletes the current job.
Why; QueueManager checks for job_container size, and if its _zero_ it knows it can perform `qrls` on this group of jobs (PBS release all jobs in this group from hold). This allows us to release groups, instead of the traditional release of all (thousands) of jobs at once, at the end.

moosebuild commented:
All [jobs](https://www.moosebuild.org/event/20834/) on afd4b94 : invalidated by @brianmoose

Only way to be sure

brianmoose commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -70,70 +77,57 @@ def __createSessionFile(self):
-                    if not os.path.exists(file):
-                        shutil.copy('../' + file, '.')
-                elif os.path.isdir('../' + file):
+                if os.path.isfile(os.path.join(source_dir, file)):
```
So many joins... it might help readability if you just have a `source_file` and `target_file` where you do the `os.path.join` once.

milljm replied:
👍

brianmoose commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -504,63 +305,267 @@ def augmentQueueParamsBase(self, job_container):
+        if json_job['std_out']:
+            stdout_file = os.path.join(json_job['working_dir'], json_job['std_out'])
+        else:
+            stdout_file=''
```
spacing

brianmoose commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -504,63 +305,267 @@ def augmentQueueParamsBase(self, job_container):
-            # no longer do we need to nerf the TestHarness's os.walk routine...)
-            self.queueJobs(status_jobs=self.__jobs)
+        if os.path.exists(stdout_file):
+            file = open(stdout_file, 'r+')
```
Is it guaranteed that the block after this if/else never throws? You might want to use a `with` on the open
just in case (which would require a bit of code restructuring).

brianmoose commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -504,63 +305,267 @@ def augmentQueueParamsBase(self, job_container):
+            # fail to find something it needed.
+            output = file.read()
+        else:
+            with self.dag_lock:
```
I see these similar 2 lines all over the place, might be worth creating a small utility method that does this

brianmoose commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -76,55 +79,56 @@ def augmentQueueParams(self, job_container, template):
-                # the results of this tester.
-                tester.setStatus(reason, tester.bucket_waiting_processing)
+                # Set the std_out path
+                job_id = re.search(r'Job Id: (\w+)', output).group(1)
```
This is guaranteed to match?

brianmoose reviewed (changes requested):
The rebuilt build boxes don't have PBS on it.
Jason noticed that the exit status isn't set correctly when PBS stuff isn't there.
https://www.moosebuild.org/job/123710/

Commit:
Adressing comments

* refactor a bunch of joins I was doing
* use a with block for testOuput
* handle qstat output a bit more elegantly

Refs #9911

Commit:
Adjust unittests for PBS

With the recent change of verifying qstat output expecting to find
both Job Id and job_state for 'valid' output, the PBS unittest needs
to be updated as well. Decided to include a real production quality
return of qstat, to better test the regexp's involved.

Refs #9911

moosebuild commented:
Job [PBS Test](https://www.moosebuild.org/job/125591/) on d0461b6 : invalidated by @milljm

we install PBS on hpcbuild10. Invalidating to try again...

moosebuild commented:
Job [Test debug](https://www.moosebuild.org/job/125582/) on d0461b6 : invalidated by @milljm

timeouts?

brianmoose reviewed (approved):
PBS is now installed on a build machine and the PBS test passes

permcody commented:
This looks ok. Changes to existing capabilities should be minimal.

permcody commented:
@milljm - I'm merging this.

milljm commented:
🎉
